edn/src/matcher.rs

```diff
@@ -0,0 +1,114 @@
+// Copyright 2016 Mozilla
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software distributed
+// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations under the License.
+
+use std::collections::HashMap;
+use regex::Regex;
+use itertools::diff_with;
+
+use types::Value;
+
+///
+pub fn match_values<T>(pattern: T, input: T) -> bool where T: AsRef<Value> {
+    use Value::*;
+    use symbols::PlainSymbol as S;
+
+    let mut named_values: HashMap<&String, &Value> = HashMap::new();
+    let name_re = Regex::new(r"^\?[^?]+$").unwrap();
+
+    match (pattern.as_ref(), input.as_ref()) {
```

Comment:
You're driving on the pattern and the input like equality, but I'd drive it from the pattern.  That is, just match the pattern, and handle the special cases (`_`, `?name`, `...`) specially.  Recurse for collection patterns and dispatch to `pattern == input` for atom patterns.  This'll be very short.